lib/lints/avoid_unnecessary_type_assertions/avoid_unnecessary_type_assertions_rule.dart

```diff
@@ -0,0 +1,183 @@
+    );
+
+    return AvoidUnnecessaryTypeAssertions._(rule);
+  }
+
+  @override
+  void run(
+    CustomLintResolver resolver,
+    ErrorReporter reporter,
+    CustomLintContext context,
+  ) {
+    context.registry.addIsExpression((node) {
+      if (_isUnnecessaryIsExpression(node)) {
+        reporter.reportErrorForNode(_unnecessaryIsCode, node);
+      }
+    });
+
+    context.registry.addMethodInvocation((node) {
+      if (_isUnnecessaryWhereType(node)) {
+        reporter.reportErrorForNode(_unnecessaryWhereTypeCode, node);
+      }
+    });
+  }
+
+  @override
+  List<Fix> getFixes() => [_UnnecessaryTypeAssertionsFix()];
+
+  bool _isUnnecessaryIsExpression(IsExpression node) {
+    final objectType = node.expression.staticType;
+    final castedType = node.type.type;
+
+    if (node.notOperator != null) {
+      if (objectType != null &&
+          objectType is! TypeParameterType &&
+          objectType is! DynamicType &&
+          !objectType.isDartCoreObject) {
+        return _isUnnecessaryTypeCheck(
+          objectType,
+          castedType,
+          isReversed: true,
+        );
+      }
+    }
+
+    return _isUnnecessaryTypeCheck(objectType, castedType);
+  }
+
+  bool _isUnnecessaryWhereType(MethodInvocation node) {
+    const whereTypeMethodName = 'whereType';
+
+    final targetType = node.target?.staticType;
+
+    if (node.methodName.name == whereTypeMethodName &&
+        isIterableOrSubclass(node.realTarget?.staticType) &&
+        targetType is ParameterizedType) {
+      final isTargetTypeHasGeneric = targetType.typeArguments.isNotEmpty;
+      final arguments = node.typeArguments?.arguments;
+      final isWhereTypeHasGeneric = arguments?.isNotEmpty ?? false;
+
+      return isTargetTypeHasGeneric &&
+          isWhereTypeHasGeneric &&
+          _isUnnecessaryTypeCheck(
+            targetType.typeArguments.first,
+            arguments?.first.type,
+          );
+    }
+    return false;
+  }
+
+  bool _isUnnecessaryTypeCheck(
+    DartType? objectType,
+    DartType? castedType, {
+    bool isReversed = false,
+  }) {
+    if (objectType == null || castedType == null) {
+      return false;
+    }
+
+    if (_isNullableCompatibility(objectType, castedType)) {
+      return false;
+    }
+
+    final objectCastedType =
+        _castedTypeInObjectTypeHierarchy(objectType, castedType);
+
+    if (objectCastedType == null) {
+      return isReversed;
+    }
+
+    if (!_isGenerics(objectCastedType, castedType)) {
+      return false;
+    }
+
+    return !isReversed;
+  }
+
+  bool _isNullableCompatibility(DartType objectType, DartType castedType) {
+    final isObjectTypeNullable = isNullableType(objectType);
+    final isCastedTypeNullable = isNullableType(castedType);
+
+    // Only one case `Type? is Type` always valid assertion case.
+    return isObjectTypeNullable && !isCastedTypeNullable;
+  }
+
+  DartType? _castedTypeInObjectTypeHierarchy(
+    DartType objectType,
+    DartType castedType,
```

Comment:
Quite a few places where these go together. Maybe a small structure will make things simpler?

```dart
class TypeCast {
  final DartType source;
  final DartType target;
}
```

Reply:
In the context of this lint, the typecast source and target are highly coupled.
In simple terms -- everywhere they go, they go together.

So if we're not using procedural/imperative style -- it looks a lot like a small struct. 

Some utility methods look like they would belong to this structure quite well -- anything that compares the two fields.